ligo/src/registryDAO.mligo

```diff
@@ -0,0 +1,190 @@
+ * 1. if proposer has locked a * s + b where s = size(pack(proposalMetadata))
+ * and a and b are from storage configuration.
+ * 2. proposal size < s_max where s_max is from stored configuration.
+ *)
+let registry_DAO_proposal_check (params, store : propose_params * storage) : bool =
+  let proposal_size = Bytes.size(Bytes.pack(params.proposal_metadata)) in
+  let frozen_scale_value = require_config ("a", store.extra) in
+  let frozen_extra_value = require_config ("b", store.extra) in
+  let max_proposal_size = require_config ("s_max", store.extra) in
+  let requred_token_lock = frozen_scale_value * proposal_size + frozen_extra_value
+  in (params.frozen_token = requred_token_lock) && (proposal_size < max_proposal_size)
+
+(*
+ * Proposal rejection return lambda: returns `c * frozen / d`. wehre c and d
+ * are specified by the DAO creator in configuration and frozen is the amount
+ * that was frozen by the proposer.
+ *)
+let registry_DAO_rejected_proposal_return_value (params, store : proposal * storage) : nat =
+  let slash_scale_value = require_config ("c", store.extra) in
+  let slash_division_value = require_config ("d", store.extra)
+  in (slash_scale_value * params.proposer_frozen_token) / slash_division_value
+
+(*
+ * Proposal rejection return lambda: returns `c * frozen / d`. where c and d
+ * are specified by the DAO creator in configuration and frozen is the amount
+ * that was frozen by the proposer.
+ *
+ * Uses the keys in the proposal metadata to differentiate between configuration
+ * proposal and normal proposals. If the map contains `agoraPostID` key, the proposal
+ * is considerd as 'normal'. Or else it is considered as carrying a 'configuration'
+ * proposal.
+ *
+ * For 'normal' proposals, the contract_extra is expected to contain the bigmap
+ * for registry and bigmap for key update tracking under keys 'registry' and
+ * 'registry_affected' keys.
+ *
+ * For 'configuration' proposal, the code expects all the configuration keys
+ * included in the proposal (or it throws an error). The expected keys are "a",
+ * "b", "s_max", "c" and "d".  'None' values are ignored and 'Some' values are
+ * used for updating corresponding configuraton.
+ *)
+let registry_DAO_decision_lambda (proposal, store : proposal * storage) : return =
+  let propose_param : propose_params = {
+    frozen_token = proposal.proposer_frozen_token;
+    proposal_metadata = proposal.metadata
+    } in
+  let proposal_key = to_proposal_key (propose_param, proposal.proposer) in
+  match Map.find_opt "agoraPostID" proposal.metadata with
+    Some (packed_b) ->
+      begin
+        match ((Bytes.unpack packed_b) : (nat option)) with
+          Some (agora_post_id) ->
+            begin
+              match Map.find_opt "updates" proposal.metadata with
+                Some (packed_b) ->
+                  begin
+                    match ((Bytes.unpack packed_b) : (((registry_key, registry_value option) map) option)) with
```

Comment:
Is it necessary to use `map` here for `updates`:

```
((registry_key, registry_value option) map)
```

AFAIU, in the spec, it is meant to be a `list` instead: https://github.com/tqtezos/baseDAO/blob/master/docs/registry.md#L25

Reply:
Good catch. Switched to use list.